### PR TITLE
Add Transceiver guide

### PR DESCRIPTION
**Pull request by mickel8**

### CODE REVIEW

**Review by LVala — COMMENTED** [review]
Some general remarks:
- use Grammarly or paste the text into ChatGPT and tell it to fix the grammar and punctuation,
- like I mentioned somewhere, I feel like the guide is missing a glossary, as a lot of the terminology might be obvious for us but not necessarily for someone who just used the JS API and didn't really look into SDP and stuff. I would also move the explanations of transceiver properties (`direction` etc.) there,
- the guide goes quite in-depth for the very beginning, I would love to see some introductory example (maybe showing differences between just using `addTrack` and using transceivers directly)
- maybe it would be nice to crate a section where all of the API functions used later on are explained (focusing on intuition)?

That's it for now, but if anything comes to my mind regarding the guide, I'm going to share it.

File: guides/transceiver_guide.md [hunk 1]
```diff
@@ -0,0 +1,512 @@
...
+* We can establish P2P connection even before obtaining access to media devices (see [Warmup](#warmup)).
+In the previous version of the API this was also possible but required creating
+a dummy track and replacing it with the real one once it was finally available.
+* They map directly to the SDP offer/answer giving high control over what is sent on which
+transceiver.
+This might have been especially important in the old days when media migh have not been bundled on
+a single ICE socket.
+In such a case, every mline could use a separate pair of ports.
+On the other hand, `addTrack` always picks the first free transceiver, which limits this control.
+* They allow for offering to receive media in a consistent to offering to send media method way.
```
[the remark this answers is not in the document]
Comment (mickel8) [reply guides/transceiver_guide.md]: I belive this is fixed now

File: guides/transceiver_guide.md [hunk 2]
```diff
@@ -0,0 +1,512 @@
...
+In the previous version of the API, user had to call `addTrack` to offer to send media and 
+`createOffer` with `{offerToReceiveVideo: 3}` to offer to only receive media, which was asymmetric and
+counter-intuitve.
+
+When speaking of transceivers there are also a couple of other notes worth mentioning before moving forward.
+* `direction` is our (local), preffered direction of the transceiver and can never be changed by applying remote offer/answer.
```
[the remark this answers is not in the document]
Comment (mickel8) [reply guides/transceiver_guide.md]: Added

File: guides/transceiver_guide.md [hunk 3]
```diff
@@ -0,0 +1,512 @@
...
+When adding a transceiver, it is by default created with `sendrecv` direction.
+When applying a remote offer that contains new mline(s), a new transceiver(s) is created with `recvonly` direction, even when the offer offers to receive media.
+* `currentDirection` is a direction negotiated between local and remote side and it changes when applying local or remote SDP
```
[the remark this answers is not in the document]
Comment (mickel8) [reply guides/transceiver_guide.md]: IMO the current one is good

**Comment by mickel8** [comment]
Some general remarks:

> use Grammarly or paste the text into ChatGPT and tell it to fix the grammar and punctuation,

done

> like I mentioned somewhere, I feel like the guide is missing a glossary, as a lot of the terminology might be obvious for us but not necessarily for someone who just used the JS API and didn't really look into SDP and stuff. I would also move the explanations of transceiver properties (direction etc.) there,
>
>the guide goes quite in-depth from the very beginning, I would love to see some introductory example (maybe showing differences between just using addTrack and using transceivers directly)
>
>maybe it would be nice to crate a section where all of the API functions used later on are explained (focusing on intuition),
>what do you think of adding a nice diagram showing relations between transceivers/senders/transports etc. with the new ex_doc diagram feature?

Great point but this is quite a lot of work and the idea was to have more advanced guide for people who more or less know what webrtc is and are looking for deep understanding of transceivers and what they can do with them.

I think that all of these things you listed i.e. glossary, more basic examples, diagrams, API explanation is for another section(s) and for another task.

For now I would favor renaming the guid to `Mastering Transceivers` or something like this and stating at the beginning what's the purpose of this guide.


> link to some introductory guide to WebRTC API in the beginning of the guide?
:+1:

File: guides/transceiver_guide.md [hunk 4]
```diff
@@ -0,0 +1,512 @@
...
+Transceiver represents an entity responsible both for sending and receiving media data.
+It consist of RTP sender and RTP receiver.
+Each transceiver maps to one mline in the SDP offer/answer.
+
+Why do we need transceivers and cannot just operate on tracks?
+* We can establish P2P connection even before obtaining access to media devices (see [Warmup](#warmup)).
+In the previous version of the API this was also possible but required creating
+a dummy track and replacing it with the real one once it was finally available.
+* They map directly to the SDP offer/answer giving high control over what is sent on which
+transceiver.
```
[the remark this answers is not in the document]
Comment (mickel8) [reply guides/transceiver_guide.md]: IMO it's okay

**Comment by LVala** [comment]
> Great point! However, this is quite a lot of work and the idea was to have more advanced guide for people who more or less know what webrtc is and are looking for deep understanding of transceivers and what they can do with them.
> 
> I think that all of the things you listed i.e. glossary, more basic examples, diagrams, API explanation is for another section(s) and for another task.
> 
> For now I would favor renaming the guide to `Mastering Transceivers` or something like this and stating at the beginning what's the purpose of this guide.

Sure

File: guides/mastering_transceivers.md [hunk 5]
```diff
@@ -0,0 +1,536 @@
...
+```
+
+<!-- tabs-close -->
+
+## Recycling m-lines
+
+When calling stop on an `RTCRtpTransceiver`, it will eventually be removed from
+the connection's set of transceivers.
+However, the number of m-lines in SDP offer/answer can never decrease.
+`m-lines` corresponding to stopped transceivers can be reused when a new transceiver appears.
```
Comment (LVala) [inline guides/mastering_transceivers.md:323]: This the only place where m-lines is written using backticks (`m-line`). Make is consistent.

File: guides/mastering_transceivers.md [hunk 6]
```diff
@@ -0,0 +1,536 @@
...
+// with index 0 and tr1 maps to m-line with index 1.
+offer = await pc1.createOffer();
+await pc1.setLocalDescription(offer);
+await pc2.setRemoteDescription(offer);
+answer = await pc2.createAnswer();
+await pc2.setLocalDescription(answer);
+await pc1.setRemoteDescription(answer);
+```
+
+### Elixir WebRTC
```
Comment (LVala) [inline guides/mastering_transceivers.md:381]: I was wondering, maybe we should just use `Elixir` for the code blocks? It's not like there's an alternative 🙂

**Review by LVala — APPROVED** [review]
Like I said earlier, maybe it would be nice to get Jellyfish guys to read through this to get and outsider's perspective, but otherwise -- sgtm.

File: guides/mastering_transceivers.md (hunk 6 above)
Comment (mickel8) [reply guides/mastering_transceivers.md:381]: Yeah I know, but Elixir WebRTC looks cool imo

**Review by LVala — COMMENTED** [review]
Firstly, I suppose it shouldn't look like this...
image

Also, I realised that, in the JS examples, you don't use `const` or `let` keywords, which is fine for console I guess, but you will need to add it when copying the examples to standalone script (at least I believe so, I'm no JS expert).

**Pull request merged**